src/OptimalFilter.jl

```diff
@@ -0,0 +1,382 @@
+                  model_params.nx+1, model_params.ny+1)
+    mu22 = model_params.obs_noise_std^(-2) * sparse(stations.ist, stations.jst,
+                                          observations,
+                                          model_params.nx+1, model_params.ny+1)
+
+    # Compute WΛWH decompositions, results are dense matrices, store them in buffers
+    # These correspond to mu21 and mu22 in Alex's code
+    WΛWH_decomposition!(offline_matrices.buf1, mu21, offline_matrices, model_params)
+    WΛWH_decomposition!(offline_matrices.buf2, mu22, offline_matrices, model_params)
+
+    # Compute the difference of the decomposition results, store in offline_matrices.buf1
+    # This corresponds to mu2 in Alex's code.
+    # TODO: Check if WΛWH is linear and you could swap the order
+    offline_matrices.buf2 .-= offline_matrices.buf1
+
+    mu10 = Vector{T}(undef, model_params.nobs)
+
+    # Loop over particles
+    for iprt = 1:filter_params.nprt
+
+        mul!(mu10, offline_matrices.R22_inv, get_values_at_stations(@view(height[:,:,iprt]), stations))
+        mu10_sparse = sparse(stations.ist, stations.jst, mu10, model_params.nx+1, model_params.ny+1)
+
+        # Compute decomposition of height values at stations times the inverse covariance matrix
+        # The argument corresponds to mu10 and the outcome to mu11 in Alex's code
+        WΛWH_decomposition!(offline_matrices.buf1, mu10_sparse, offline_matrices, model_params)
+
+        # Compute the mean for the ith particle using mu2 and mu11
+        # Skip storing the temporary mu1 in Alex's code
+        mean[:,:,iprt] .= @view(height[:,:,iprt]) .- offline_matrices.buf1 .+ offline_matrices.buf2
+
+    end
+
+end
+
+function sample_height_proposal!(height::AbstractArray{T,3},
+                                 offline_matrices::OfflineMatrices, online_matrices::OnlineMatrices,
+                                 observations::AbstractVector{T}, stations, model_params, filter_params,
+                                 rng::Random.AbstractRNG) where T
+
+    @assert filter_params.nprt % 2 == 0 "Number of particles must be even"
+
+    calculate_mean_height!(online_matrices.mean, height, offline_matrices, observations, stations, model_params, filter_params)
+
+    #TODO: Can you create a Normal(0,1) distribution of complex type?
+    nd = Normal(0,1)
+
+    i_n1 = LinearIndices((model_params.nx+1, model_params.ny+1))
+    i_n1_bar = LinearIndices((2*model_params.nx, 2*model_params.ny))
+
+    for iprt in 1:2:filter_params.nprt
+        # TODO randn(ComplexF64) seems to set variance = 1/2. Could not find how to change that.
+        # TODO we could pre-create all our random numbers in one go before the loop, would that be faster?
+        e1 = rand(rng, nd, 4*model_params.nx*model_params.ny) .+ rand(rng, nd, 4*model_params.nx*model_params.ny)im
+        e2 = rand(rng, nd, model_params.nobs) .+ rand(rng, nd, model_params.nobs)im
```

Comment:
This doesn't need a special distribution:
```suggestion
        e1 = complex.(randn(rng, 4*model_params.nx*model_params.ny), randn(rng,4*model_params.nx*model_params.ny))
        e2 = complex.(randn(rng, model_params.nobs), randn(rng, model_params.nobs))
```

Reply:
This change breaks the validation test. Any idea why? I'm guessing the random stream gets screwed up somehow but I don't see how

Reply:
It's not unlikely that the stream of numbers generated by `rand(Normal(1, 0), ...)` is different from `randn(...)`, but their statistical properties are the same

Reply:
Actually, the stream seems to be the same:
```julia
julia> using Random, Distributions

julia> rng = Random.seed!(42)
MersenneTwister(42)

julia> randn(rng, 10)
10-element Vector{Float64}:
 -0.5560268761463861
 -0.444383357109696
  0.027155338009193845
 -0.29948409035891055
  1.7778610980573246
 -1.14490153172882
 -0.46860588216767457
  0.15614346264074028
 -2.641991008076796
  1.0033099014594844

julia> rng = Random.seed!(42)
MersenneTwister(42)

julia> rand(rng, Normal(0, 1), 10)
10-element Vector{Float64}:
 -0.5560268761463861
 -0.444383357109696
  0.027155338009193845
 -0.29948409035891055
  1.7778610980573246
 -1.14490153172882
 -0.46860588216767457
  0.15614346264074028
 -2.641991008076796
  1.0033099014594844
```
:thinking:

Reply:
This is how sampling of `Normal(μ, σ)` is implemented: https://github.com/JuliaStats/Distributions.jl/blob/863844c88e4153af13996f571fcc612d159de542/src/univariate/continuous/normal.jl#L243, so yeah, there is no reason why it should be different :confused:

Reply:
Yeah, I checked the same. Oddly enough in the validation test it does not work.

If I change both sides of the test to use `randn` then it works again 🤔